First draft for domain-based application testing (#100)
mulesoft/docs-munit: modules/ROOT/pages/domain-based-application-cookbook.adoc

= Test Domain-Based Applications
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> MUnit 2.2 allows you to write MUnit test suites for your Mule applications that share their resources with a Mule domain without the need for any extra configurations.

Reviewer: Maybe we should change this to 2.x instead of 2.2, since the feature is present in any version of MUnit 2, and otherwise we would need to change this each time a new version comes out.

== Base Mule Application

Consider a Mule domain called `domain-config` that defines a global configuration for the HTTP listeners of all the Mule applications under it:

.Mule domain
[source,XML,linenums]
----
<domain:mule-domain ...>

<http:listener-config name="HTTP_Listener_config" basePath="/" >
  <http:listener-connection host="0.0.0.0" port="8081" />
</http:listener-config>

</domain:mule-domain>
----

.Mule domain pom.xml file
[source,XML,linenums]
----
<groupId>com.mycompany</groupId>
<artifactId>domain-config</artifactId>
<version>1.0.0-SNAPSHOT</version>
<packaging>mule-domain</packaging>
<name>domain-config</name>

----

And a basic Mule application that returns the path being pinged:

[source,XML,linenums]
----
<mule ...>

  <flow name="application-aFlow">
        <http:listener  config-ref="HTTP_Listener_config" path="/api"/>
        <set-payload value='#["You reached " ++ message.attributes.listenerPath]'/>
  </flow>

> </mule>

Reviewer: The indentation in this snippet does not match the rest of the snippets. This makes it look out of place.

[source,XML,linenums]
----
<dependencies>
...
  <dependency>
    <groupId>com.mycompany</groupId>
    <artifactId>domain-config</artifactId>
    <version>1.0.0</version>
    <classifier>mule-domain</classifier>
    <scope>provided</scope>
  </dependency>
...
</dependencies>
----

== Write The MUnit Test

Because the Mule domain is a dependency of the Mule application, MUnit can start up the Mule domain to pick up the necessary global elements configuration for the application that it's testing.

Assume an MUnit test to validate that the path being pinged is indeed the expected `/api` path:

[source,XML,linenums]
----

<mule ...>

  <munit:config name="application-a-test-suite.xml" />

  <http:request-config name="HTTP_Request_configuration" basePath="/api">
    <http:request-connection host="0.0.0.0" port="8081" />
  </http:request-config>

  <munit:test name="application-a-test-suite-application-aFlowTest" description="Test to validate the path being called">

  <munit:enable-flow-sources >
    <munit:enable-flow-source value="application-aFlow" />
  </munit:enable-flow-sources>

  <munit:execution>

>     <http:request config-ref="HTTP_Request_configuration" method="GET"/> //<1>

Reviewer: This //<1> is unnecessary since there are no details after the snippet. Same goes for //<2> and //<3>

Author: We are using numbered lists for these callaouts because the style for the proper `<1>` is not entirely formated to match our style.
These are followed by:

```
. Inside the `execution` scope, an HTTP requester sends a `GET` request to your application's endpoint.
. Inside the `validation` scope, set up an `assert-equals` processor to assert the response from the `execution` scope against your expected response.
. To be able to compare the response against the expected text you must convert the payload to plain text.
```

    <ee:transform>
      <ee:message>
        <ee:set-payload> //<2>
          <![CDATA[%dw 2.0
          output text/plain
          ---
          payload
          ]]>
        </ee:set-payload>
      </ee:message>

>     </ee:transform>

Reviewer: There should not be a transform message in the execution lane. Should be moved to validation or be merged with the assertion as follows: `<munit-tools:assert-equals actual="#[output text/plain --- payload]" expected='#["You reached /api"]'/>`

> = Test Domain-Based Applications

Reviewer 2: This topic set is really an extended example illustrating how to write a test suite for an example domain-based application by providing example values and then using them in steps explaining how to write a test for that example.

So I think we need a title and a topic introduction that tells under what circumstances a user might want or need to perform such a test, and then mention that the process is most easily explained by providing an example environment and using it to illustrate the steps required to write?create?perform? the test.

I'm not sure offhand how to word a more specific title, or even if that is necessary, since this is about testing (an example) domain-based application. Hmm... How to Write an MUnit Test Suite for Domain-Based Applications? Example of Writing an MUnit Test Suite....?

But I will try to offer some ideas for text around the two subtopics.

ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> MUnit 2.x allows you to write MUnit test suites for your Mule applications that share their resources with a Mule domain without the need for any extra configurations.

Reviewer 2: ```suggestion
MUnit 2.x enables you to write MUnit test suites for your Mule applications that share their resources with a Mule domain, without you performing any additional configuration. To understand how this works, it is helpful to consider a specific scenario populated with actual values, and then use those values in steps that illustrate how to write such a test.
```

> == Base Mule Application

Reviewer 2: ```suggestion
== Example Domain-Based Mule Application to Be Tested
```
Does that sound OK?

> Consider a Mule domain called `domain-config` that defines a global configuration for the HTTP listeners of all the Mule applications under it:

Reviewer 2: ```suggestion
Consider a Mule domain called `domain-config` that defines a global configuration for the HTTP listeners of all the Mule applications within it:
```
I think things are "in" a domain rather than "under" it?

[.ex]
=====
[discrete.view]

> ===== Studio Visual Editor

Reviewer 2: I'm not sure if it is clearer to simply provide these internal headings with their examples, or to make these introductory sentences to the examples, such as "The Anypoint Studio Visual Editor configuration looks like this:"
Just something to think about, for this ===== and the others below.

image::domain-munit-example.png[]

[discrete.view]
===== XML Editor

.Mule domain
[source,XML,linenums]
----
<domain:mule-domain ...>

<http:listener-config name="HTTP_Listener_config_domain" basePath="/" >
  <http:listener-connection host="0.0.0.0" port="8081" />
</http:listener-config>

</domain:mule-domain>
----

.Mule domain pom.xml file
[source,XML,linenums]
----
<groupId>com.mycompany</groupId>
<artifactId>domain-config</artifactId>
<version>1.0.0-SNAPSHOT</version>
<packaging>mule-domain</packaging>
<name>domain-config</name>
----
=====

> And a basic Mule application that returns the path being pinged:

Reviewer 2: ```suggestion
Consider also a basic Mule application that returns the path being pinged:
```

[.ex]
=====
[discrete.view]
===== Studio Visual Editor

image::mule-app-domain-listener.png[]

> Note that this application does not define a HTTP Listener global configuration because it inherits it from its Mule domain (`HTTP_Listener_config_domain`).

Reviewer 2: Listener > listener

lowercased above and in most Google searches....

image::mule-app-domain-set-payload.png[]

[discrete.view]
===== XML Editor

[source,XML,linenums]
----
<mule ...>

  <flow name="application-aFlow">
        <http:listener  config-ref="HTTP_Listener_config_domain" path="/api"/>
        <set-payload value='#["You reached " ++ message.attributes.listenerPath]'/>
  </flow>

</mule>
----

> Note that this application does not define an `http:listener-config` global element because it inherits it from its Mule domain (`HTTP_Listener_config_domain`). The application declares the domain dependency in its `pom.xml` file:

Reviewer 2: Make this sentence and Line 56 exactly the same, one way or the other. 
Avoid saying one thing two (or more) different ways.

Author: This is meant for XML users as opposite from the Studio users.
In the past, we had different tabs that users can select, but that was not supported in the current UI and was migrated to this structure.

[source,XML,linenums]
----
<dependencies>
...
  <dependency>
    <groupId>com.mycompany</groupId>
    <artifactId>domain-config</artifactId>
    <version>1.0.0</version>
    <classifier>mule-domain</classifier>
    <scope>provided</scope>
  </dependency>
...
</dependencies>
----

=====

> == Write The MUnit Test

Reviewer 2: ```suggestion
== Writing an MUnit Test of the Example Domain-Based Application
```

> Because the Mule domain is a dependency of the Mule application, MUnit can start up the Mule domain to pick up the necessary global elements configuration for the application that it's testing.

Reviewer 2: ```suggestion
Because the Mule domain is a dependency of the Mule application, MUnit can start the Mule domain to retrieve the necessary global elements configuration for the application that it's testing.
```

Assume an MUnit test to validate that the path being pinged is indeed the expected `/api` path:

[.ex]
=====
[discrete.view]

> ===== Studio Visual Editor

Reviewer 2: ditto previous comment about these ===== headings vs. intro sentences

image::munit-domain-sample.png[]

. Inside the `execution` scope, an HTTP requester sends a `GET` request to your application's endpoint.
. Inside the `validation` scope, set up an Assert Equals` processor to assert the response from the `execution` scope against your expected response.

> . To be able to compare the response against the expected text you must convert the payload to plain text.

Reviewer 2: These are formatted as either steps in a task or a chronological list of events, but they are worded both ways.
Can we make them parallel? 
I think the easiest way to do this is to slightly reword #2 to something like 

"Inside the `validation` scope, you decide to set up an Assert Equals` processor to assert the response from the `execution` scope against your expected response."

Something that does not use imperative voice (do this).

Also, "To be able to compare the response against the expected text you must..."  >  "To compare the response against the expected text, you must..."

Author: These are not steps. These are callouts from the code callouts, or from the callouts in the screenshots.
Ideally, we should use <1>, instead of numbered lists, but we need to fix the style of the site for those to display properly.

[discrete.view]
===== XML Editor

[source,XML,linenums]
----

<mule ...>

  <munit:config name="application-a-test-suite.xml" />

  <http:request-config name="HTTP_Request_configuration" basePath="/api">
    <http:request-connection host="0.0.0.0" port="8081" />
  </http:request-config>

  <munit:test name="application-a-test-suite-application-aFlowTest" description="Test to validate the path being called">

  <munit:enable-flow-sources >
    <munit:enable-flow-source value="application-aFlow" />
  </munit:enable-flow-sources>

  <munit:execution>
    <http:request config-ref="HTTP_Request_configuration" method="GET"/> //<1>
  </munit:execution>

  <munit:validation>
    <munit-tools:assert-equals //<2>
        actual="#[output text/plain --- payload]" //<3>
        expected='#["You reached /api"]'/>
  </munit:validation>

</munit:test>
----

. Inside the `execution` scope, an HTTP requester sends a `GET` request to your application's endpoint.

> . Inside the `validation` scope, set up an `assert-equals` processor to assert the response from the `execution` scope against your expected response.

Reviewer 2: ditto
```suggestion
. Inside the `validation` scope, you decide to set up an `assert-equals` processor to assert the response from the `execution` scope against your expected response.
```

> . To be able to compare the response against the expected text you must convert the payload to plain text.

Reviewer 2: ```suggestion
. To compare the response against the expected text, you must convert the payload to plain text.
```
